Compile time annotation processor based SVC architecture

skydoves opened the pull request:
### Description
안녕하세요~
지난번 #6 에 이어서 새로운 PR 입니다.

기존의 문제점은 아래와 같습니다.
1. Generic을 사용하면 type을 명확하게 지정해 줄 수 있는 장점이 있는 반면, 복잡한 상속의 구조 때문에 
개발 및 유지보수에 상당히 복잡함과 햇갈림이 우려 될 수 있다고 생각했습니다.

2. #6 의 대체 방안은 annotation을 활용한 Generic의 제거와, 사용자 기준의 단순화였습니다.
하지만 이 또한 Generic을 제거하였으나 명확한 type구조를 ControwlTower에서 알 수 없기 때문에 screen과 view에 대한 field들을 생성하고 캐스팅을 해주어야 하고, 조금 더 세련된 방식인 [dependency injection](https://github.com/naver/svc/pull/6/commits/7fec9e6351fc5566573a4312abd523bb34b60e85)을 고안했으나 ConotrolTower에서 field를 정의해야 하는 부분은 여전이 이슈가 되었습니다.

3. 이번에 새롭게 수정된 apt 방식의 아키텍쳐입니다. 기존의 `svc-svc` 모듈의 qualifier들을 `svc-annotation`으로 분리시켰으며, `svc-compiler`를 통해서 runtime이 아닌 compile-time에  `getScreen` 메서드와 `getViews` 메서드의 리턴타입이 지정된 ControlTower class를 생성시켜주는 방식입니다. 이렇게 되면 `ControlTower.class` 를 상속받고 있는 ControlTower class를 빌드 후, 다시 새롭게 generated된 `SVC-ControlTower` class로 상속을 바꿔주어야 한다는 단점이 있습니다. 대신, 기존에 screen과 view 두가지 필드를 선언할 필요도 없고, 직접 개발자가 정의해서 캐스팅이나 di를 통하여 타입 변환을 시켜주지 않아도 됩니다.

=> 한가지 대체 방안은, ControlTower class의 이름을 ActivityName + ControlTower와 같이 규제를 한다면 (혹은 template를 활용) `@RequireControlTower(ControlTower.class)` annotation 내부에 class value 없이도 ControlTower class의 full name을 알기 때문에 instance 생성이 가능해집니다. 따라서 Activity생성시 value없이 `@RequireControlTower` annoation만 붙이고 build 후, generated된 `SVC-ControlTower`를 상속받는 ControlTower class를 작성할 수 있습니다.

### Code difference
#### Before
```kotlin
lass StatisticControlTower : ControlTower(), StatisticViewsAction {

    @InjectView
    lateinit var statisticViews: StatisticViews

    override fun onNameClicked() {
        showToast("NameClicked")
    }

    override fun onCreated() {
        this.statisticViews.setName(this.statisticViews.javaClass.simpleName)
        this.statisticViews.setExtra(javaClass.simpleName)
    }
}
```

#### After
```kotlin
class StatisticControlTower : SVC_StatisticFragmentControlTower(), StatisticViewsAction {

    override fun onNameClicked() {
        showToast("NameClicked")
    }

    override fun onCreated() {
        view.setName(view.javaClass.simpleName)
        view.setExtra(javaClass.simpleName)
    }
}
```

### What happened?
Here is an example of `MainActivity.class` with annotated `@RequireControlTower` annotation.
#### MainActivity.class
```kotlin
@RequireControlTower(MainControlTower::class)
class MainActivity : SvcActivity<MainViews>() {
  // -- skip --
}
```
#### MainControlTower.class
```kotlin
class MainControlTower : ControlTower(), MainViewsAction {
  // -- skip --
}
```
Annotate `@RequireControlTower` annotation and __build__ your project.
After build success, `SVC_(Activity Name)ControlTower` class will be generated automatically by svc-compiler. Then re-write the extending class as the `SVC_(Activity Name)ControlTower` class.
#### *MainControlTower.class
```kotlin
class MainControlTower : SVC_MainActivityControlTower(), MainViewsAction {
  override fun onCreated() {
        view.setName(view.javaClass.simpleName)
        view.setExtra(javaClass.simpleName)
    }
}
```

### Changed
- added svc-annotation, svc-compiler modules
- implements svc-compiler for generating SVC_ControlTower class
- removed dependency injection using annotation
- removed @InjectScreen @InjectView qualifiers

apt 아키텍쳐의 큰 틀입니다. 여유로우실때 리뷰 부탁드립니다. :) 👍

BansookNam commented:
Wow. 👍 
늦은시간 까지 고생많으셨고, 조만간 확인하도록 하겠습니다~!

BansookNam commented on svc/src/main/kotlin/com/naver/android/svc/core/screen/SvcActivity.kt:
```diff
@@ -36,7 +36,6 @@ import com.naver.android.svc.core.views.Views
 @Suppress("PrivatePropertyName")
 abstract class SvcActivity<out V : Views> : AppCompatActivity(), Screen<V>, DialogPlug {
```
최신 AppCompatActivity가 setContentView에서
windows null 여부를 체크하더라고요.

SvcActivity. onCreate에서 super.onCreate를 가장 먼저 
호출하도록 수정이 필요합니다.

skydoves replied:
감사합니다.
수정되었습니다!

BansookNam commented on svc/src/main/kotlin/com/naver/android/svc/core/controltower/ActivityControlTowerManager.kt:
```diff
@@ -19,26 +19,32 @@ class ActivityControlTowerManager {
     fun <T : ControlTower> fetch(activity: SvcActivity<*>,
-                                 controlTowerClass: KClass<T>,
+                                 controlTowerClass: KClass<*>,
                                  views: Views): T {
```
1. ControlTowerManager 하나로 합치기 의견
ActivityControlTowerManager, DialogFragmentControlTowerManager, FragmentControlTowerManager
하는 역할이 거의 같은데, ControlTowerManager로 합치고
fun <T : ControlTower> fetch(screen: Screen<*>,
                                 controlTowerClass: KClass<T>,
                                 controlTowerClass: KClass<*>,
                                 views: Views): T
스크린을 받도록 바꾸는건 어떨까요?

2. Manager안에 HashMap으로 ControlTower들을 가지고 있는 이유는 어떤걸까요?
    어디서든 필요한 Contower에 쉽게 접근할 수 있게 하려는건가요?

    보통 다른 화면에 뭔가 notify를 할때 EventBus를 쓰거나
    ActivityScope의 ViewModel을 쓰곤 했었는데,
    ControlTowerManager를 통해서 직접 접근해서 수행하는 것도 좋겠네요.
    기존에 Fragment에 접근할 일이 있으면 FragmentManager통해서 힘들게 찾던 일을
    ControlTowerManager을 통하면 좀더 쉬울거 같고요.
    제가 말한게 맞는지요??

3. APT 대박입니다! 고생 많으셨습니다. 이제 딱 하나 남았습니다.
   마지막으로 각 screen에서 views는 제네릭 덕분에 접근이 가능한데,
   controlTower는 캐스팅을 해주어야합니다.
  이것도 apt로 가능하려는지요? SvcFragment를 상속하지 않고 SVC_HomeFragment, SVC_MainActivity
  이렇게 생성되면 가능할거 같은데요. 가능한지 한번 살펴봐주세요.

3번이 해결되면, 머지하고, 제가 조금더 손 본 뒤에 
master에 머지하도록 하겠습니다.
예상했던것보다 훨씬 빨리 끝내주셔서, 4월 5일에는 annotation기반으로 발표해도 될거같아요.
감사합니다.
    

의견 주시면 감사하겠습니다.

skydoves replied:
1. 저도 동의하는 부분입니다. 이번 커밋에서 전부 동일한 `ControlTowerManager` class로 통합되었습니다.
2. 맞습니다. 하지만 ControlTower를 create 및 destroy하고 전체적으로 생명주기를 관리하는 부분에 조금 더 의미를 두고 있습니다.
3.  감사합니다 ㅎㅎ Activity에서 controlTower접근 캐스팅도 이번 커밋에서 전부 해결 될 예정입니다!

BansookNam commented on svc-compiler/src/main/java/com/naver/android/compiler/SvcProcessor.java:
```diff
@@ -0,0 +1,106 @@
+
+  private void checkTypeValidation(TypeElement annotatedType) throws IllegalAccessException {
+    if (!annotatedType.getKind().isClass()) {
+      throw new IllegalAccessException("Only classes can be annotated with @PreferenceRoom");
```
PreferenceRoom -> RequireControlTower 수정 부탁드립니다.

skydoves replied:
감사합니다~
수정되었습니다!

BansookNam commented:
계속 보다보니 욕심이 생기네요.

SVC_MainActivity 등으로 apt가 된다면

@RequireControlTower(StatisticControlTower::class)
-> @ControlTower(StatisticControlTower::class)
이름 변경을 하고

@Views(HomeViews::class)

Views도 annotation기반이되면, controlTower랑 views둘다 공통 문법으로 쓰니
보기도 좋을것 같습니다. 너무 무리하진 마시고, 검토 부탁드립니다.

skydoves commented:
### Description
두번째 커밋입니다!
리뷰에서 제안해주신 부분이 거의다 반영되었으며 새롭게 추가된 부분도 몇가지 있습니다.

`SVC_Activity`, `SVC_fragment`를 compile-time에 auto generating하여 상속 받으면 `controlTower()` 메서드의 리턴타입이 지정된 Activity 혹은 Fragment class를 생성시켜주는 방식입니다. 

 `@RequireControlTower`만을 사용해서 `SVC_Activity` class를 generate하는 방법에는 몇가지 오류가 있습니다.

#### 문제1
한가지 문제점은  `@RequireControlTower` annoation이 붙은 class에 대하여 superClass(`SvcActivity<MainViews>`) 정보를 참조하여 `SVC_Activity` class를 generate 했는데, 생성된 `SVC_Activty`를 새롭게 상속받으면, 다음 compile-time때는 `SVC_Activity`를 superClass로 상속 받는 이상한(?) class가 generate 되는 재귀 문제가 있습니다.

#### 문제2
 compile-time에 class의 데이터를 읽기 전에 기존에 generated된 class들은 전부 삭제되고 새로 생성한다는 부분입니다. 따라서 새롭게  `SVC_Activity`를 상속받고 있어도 해당 클래스를 참조 할 수 없습니다.

#### 해결방안
 따라서 위와 같은 문제를 해결하기 위해 `@ControlTower`라는 annotation을 만들었고, `@ControlTower` annotation의 type element 정보를 가져오는 scope에서 `SVC_Activity`를 generate하도록 개발 되었습니다. 또한 새롭게 생성되어야 하는  `SVC_Activity`의 superClass type을 지정해주기 위해 `OwnSvcActivity`, `OwnSvcFragment`, `OwnSvcDialogFragment` annotation들이 추가 되었습니다.

### What happened?
#### Before
MainActivity.class
```kotlin
@RequireControlTower(MainControlTower::class)
class MainActivity : SvcActivity<MainViews>() {
  // -- skip --
}
```
MainControlTower.class
```kotlin
class MainControlTower : SVC_MainActivityControlTower(), MainViewsAction {
  // -- skip --
}
```

#### After
MainActivity.class
```kotlin
@OwnSvcActivity
@RequireViews(MainViews::class)
@RequireControlTower(MainControlTower::class)
class MainActivity : SVC_MainActivity() {
  // -- skip --
}
```
MainControlTower.class
```kotlin
@ControlTower
class MainControlTower : SVC_MainActivityControlTower(), MainViewsAction {
  // -- skip --
}
```

`@RequireControlTower` => generates `SVC_MainActivityControlTower` class on compile-time.
`@ControlTower` => generates `SVC_MainActivity ` class on compile-time.

- integrated ActivityControlTowerManager, FragmentControlTowermanager, DialogFragmentControlTowerManager to ControlTowerManager
- added `@ControlTower`, `@OwnSvcActivity', `@OwnSvcFragment`, `@OwnSvcDialogFragment` annotations.
- implemented generating `SVC_Activity` ~ `SVC_Fragment` ~ `SVC_DialogFragment`
- rearrange & optimized all codes

저도 생각보다 작업이 일찍 끝나서 다행인 것 같습니다~
요즘 또 날씨가 추워졌는데 감기 조심하시구, 천천히 리뷰 부탁드립니다! :)

BansookNam commented:
이야기한걸 다 이루셨네요 ㅎㅎ 대단하십니다. 
고생많으셨구요, 네이밍 등 몇가지 수정할게 있는데.
먼저 머지하고 제가 수정해놓겠습니다.

발표준비 마무리를 해야되서 제 커밋은 PR로 커뮤니케이션하기가 
시간이 부족하여, 바로바로 PR을 직접 머지하도록 하겠습니다. 
혹시 이슈될만한게 커밋에 있으면 코멘트 부탁드립니다!

skydoves commented:
@BansookNam 
리뷰하시느라 고생하셨습니다~
저도 계속 모니터링하면서 런칭까지 필요한 부분이 생기면 또 채워넣도록 하겠습니다.ㅎㅎ
그리고 발표전까지 제가 도와드릴 수 있는 부분이 있다면 언제든지 도와드리겠습니다. :)
고생 많으셨습니다! 👍
